src/withStateHandlers.js

```diff
@@ -0,0 +1,77 @@
+import createHelper from './createHelper'
+import callOrUse from './utils/callOrUse'
+import updateProps from './utils/updateProps'
+import mapValues from './utils/mapValues'
+
+/**
+ * Passes state object properties and immutable updater functions in a form of
+ * `(...payload: any[]) => Object` to the base component.
+ *
+ * Every state updater function accepts state, props and payload and must return
+ * a new state or undefined. The new state is shallowly merged with the previous
+ * state.
+ *
+ * @static
+ * @category Higher-order-components
+ * @param {Object|Function} initialState
+ * @param {Object} stateUpdaters
+ * @returns {HigherOrderComponent} A function that takes a component and returns a new component.
+ * @example
+ *
+ * const Counter = withStateHandlers(
+ *   ({ initialCounter = 0 }) => ({
+ *     counter: initialCounter,
+ *   }),
+ *   {
+ *     incrementOn: ({ counter }) => (value) => ({
+ *       counter: counter + value,
+ *     }),
+ *     decrementOn: ({ counter }) => (value) => ({
+ *       counter: counter - value,
+ *     }),
+ *     resetCounter: (_, { initialCounter = 0 }) => () => ({
+ *       counter: initialCounter,
+ *     }),
+ *   }
+ * )(
+ *   ({ counter, incrementOn, decrementOn, resetCounter }) =>
+ *     <div>
+ *       <Button onClick={() => incrementOn(2)}>Inc</Button>
+ *       <Button onClick={() => decrementOn(3)}>Dec</Button>
+ *       <Button onClick={resetCounter}>Reset</Button>
+ *     </div>
+ * )
+ */
+const withStateHandlers = (initialState, stateUpdaters) =>
+  updateProps(next => {
+    let props
+    let state
+
+    const handlers = mapValues(
+      stateUpdaters,
+      handler => (mayBeEvent, ...args) => {
+        if (mayBeEvent && typeof mayBeEvent.persist === 'function') {
```

Comment:
@matthieuprat sure, updated to call `persist()` on `SyntheticEvent` like Recompose and added corresponding test (from Recompose), the only reason I didn't include this originally is that we don't call `setState()` (using callback) like Recompose so I didn't know if it was necessary, but without fully understanding `recompact` internals it seems safer to `persist()`

Reply:
Oh, you're right, I misread the PR description. Indeed, we don't need to persist the event since we're always calling the handler synchronously. Sorry for the confusion.